source/extensions/quic_listeners/quiche/envoy_quic_proof_source.cc

```diff
@@ -0,0 +1,103 @@
+#include "extensions/quic_listeners/quiche/envoy_quic_proof_source.h"
+
+#include <openssl/bio.h>
+
+#include "envoy/ssl/tls_certificate_config.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+#include "extensions/quic_listeners/quiche/quic_io_handle_wrapper.h"
+#include "extensions/transport_sockets/well_known_names.h"
+
+#include "openssl/bytestring.h"
+#include "quiche/quic/core/crypto/certificate_view.h"
+
+namespace Envoy {
+namespace Quic {
+
+quic::QuicReferenceCountedPointer<quic::ProofSource::Chain>
+EnvoyQuicProofSource::GetCertChain(const quic::QuicSocketAddress& server_address,
+                                   const quic::QuicSocketAddress& client_address,
+                                   const std::string& hostname) {
+  absl::optional<std::reference_wrapper<const Envoy::Ssl::TlsCertificateConfig>> cert_config_ref =
+      getTlsCertConfig(server_address, client_address, hostname);
+  // Only return the first TLS cert config.
+  if (!cert_config_ref.has_value()) {
+    ENVOY_LOG(warn, "No matching filter chain found for handshake.");
+    return quic::QuicReferenceCountedPointer<quic::ProofSource::Chain>(
+        new quic::ProofSource::Chain({}));
+  }
+  auto& cert_config = cert_config_ref.value().get();
+  const std::string& chain_str = cert_config.certificateChain();
+  std::string pem_str = std::string(const_cast<char*>(chain_str.data()), chain_str.size());
+  std::stringstream pem_stream(chain_str);
+  std::vector<std::string> chain = quic::CertificateView::LoadPemFromStream(&pem_stream);
+  if (chain.empty()) {
+    throw EnvoyException(
```

Comment:
I think it's not safe to throw exception because this function is called from shared quic code, e.g. `QuicCryptoServerConfig::EvaluateClientHello`, which assumes no exception is thrown.

Can we return a nullptr instead?

Reply:
return empty Chain here instead.